File: app/src/main/java/com/verbatoria/business/session/activities/UserInteractionTimerTask.java
package com.verbatoria.business.session.activities;

import android.content.Context;
import com.verbatoria.VerbatoriaApplication;
import com.verbatoria.presentation.login.view.sms.SMSConfirmationActivity;
import javax.inject.Inject;

/**
 * Таймер для неактивности - через 30 минут стартуем смс подтверждение
 *
 * @author nikitaremnev
 */
public class UserInteractionTimerTask {

    //60-minutes, 1 hour
    private static final long USER_INTERACTION_INTERVAL = 1000 * 60 * 60;

    @Inject
    public Context mContext;

    private long lastInteractionTime = System.currentTimeMillis();
    private boolean isSmsConfirmationStarted = false;

    public UserInteractionTimerTask() {
        VerbatoriaApplication.getInjector().inject(this);
    }

    public void updateLastInteractionTime() {
        if (System.currentTimeMillis() - lastInteractionTime > USER_INTERACTION_INTERVAL) {
            startSMSConfirmation();
        } else {
            lastInteractionTime = System.currentTimeMillis();
        }
    }

    public void dropTimerTaskState() {
        isSmsConfirmationStarted = false;
        lastInteractionTime = System.currentTimeMillis();
    }

    private void startSMSConfirmation() {
        if (!isSmsConfirmationStarted) {
            mContext.startActivity(SMSConfirmationActivity.Companion.newInstance(mContext));
        }
        isSmsConfirmationStarted = true;
    }

}

File: app/src/main/java/com/verbatoria/domain/authorization/manager/UserInteractionTimerTask.kt
package com.verbatoria.domain.authorization.manager

import android.content.Context
import com.verbatoria.VerbatoriaKtApplication
import com.verbatoria.domain.authorization.repository.AuthorizationRepository
import com.verbatoria.infrastructure.extensions.MILLISECONDS_IN_SECOND
import com.verbatoria.ui.login.sms.SMSLoginActivity
import com.verbatoria.utils.CountryHelper
import javax.inject.Inject

/**
 * Таймер для неактивности - через 30 минут стартуем смс подтверждение
 *
 * @author nikitaremnev
 */

private const val USER_INTERACTION_INTERVAL = (MILLISECONDS_IN_SECOND * 60 * 60).toLong()

class UserInteractionTimerTask {

    @Inject
    lateinit var context: Context

    @Inject
    lateinit var authorizationRepository: AuthorizationRepository

    private var lastInteractionTime = System.currentTimeMillis()
    private var isSmsConfirmationStarted = false
    private val currentCountry: String

    init {
        VerbatoriaKtApplication.injector.inject(this)
        currentCountry = authorizationRepository.getCurrentCountry()
    }

    fun updateLastInteractionTime() {
        if (!CountryHelper.isCountryRequireSkipSMSConfirmation(currentCountry) && System.currentTimeMillis() - lastInteractionTime > USER_INTERACTION_INTERVAL) {
            startSMSConfirmation()
        } else {
            lastInteractionTime = System.currentTimeMillis()
        }
    }

    fun dropTimerTaskState() {
        isSmsConfirmationStarted = false
        lastInteractionTime = System.currentTimeMillis()
    }

    private fun startSMSConfirmation() {
        if (!isSmsConfirmationStarted) {
            context.startActivity(SMSLoginActivity.newInstance(context))
        }
        isSmsConfirmationStarted = true
    }

}